### DISCUSSION TITLE
Hi I wanted to ask if it is possible to read a `.nwk` file with msprime or tskit

### DISCUSSION THREAD

**Discussion by AyomideOjo** [tskit-dev/msprime#2191] [General]
Hi I wanted to ask if it is possible to read a `.nwk` file with msprime or tskit

**Comment by benjeffery** [comment]
Hi @AyomideOjo, there are some methods in tsconvert that may be of use. https://github.com/tskit-dev/tsconvert

**Comment by AyomideOjo** [comment]
Good day,
If you could please direct me to those fuctions that wouls would be amazing,  because i am still having problems.
Thank-you
Ayomide

Envoyé à partir de Outlook pour Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by AyomideOjo** [comment]
sorry I just clicked the link you sent, i found that after i made the post,
but i am having some problems with working with self.kc_distance(tree)
afterwards.
thank you
Ayomide

Le jeu. 3 août 2023, à 05 h 14, Ben Jeffery ***@***.***> a
écrit :

**Comment by jeromekelleher** [comment]
Hi @AyomideOjo :wave: 

Perhaps you could give us a little background on what you're trying to do and we might be able to point you in the right direction?

**Comment by AyomideOjo** [comment]
Hello, I'm a McGill undergraduate working in a lab over the summer. I was asked to translate `kc_distance` from R to Python, where the tree data is provided from a .nwk file. It has been a surprisingly difficult task. Initially, I attempted to solve the problem using Multiprocessing and Cython; however, that approach didn't work out for trees of large sizes. Eventually, I found 'tskit' and was able to download it. However, the library function `kc_distance` only works with `tskit.treesequence` data.

I then tried converting the Newick trees to `tskit.treesequence` trees, but I encountered some problems. Firstly, the data I received had branch lengths of 0.0, which the `from_newick()` function rejected. To resolve this, I changed the `min_edge_length` variable to be equal to `sys.float_info.epsilon * 1.25`.

Another error I stumbled upon was that the `tskit.treesequence` tree generated by the function seems to have nodes labeled with IDs given by the function `add_row`. Furthermore, the function `kc_distance` appears to call another function named `check_kc_distance_samples_inputs` to verify that the tree's tips have the same names. However, my error here is that the `samples` attribute returns a list of IDs, which do not match the names of the tips in the data I was provided with. I presume that the `add_row` function assigns IDs incrementally, starting with 0, then 1, then 2, and so on. Consequently, the for loop in the function `check_kc_distance_samples_inputs` detects that two leaf nodes have different IDs, leading to the error `TSK_ERR_SAMPLES_NOT_EQUAL`. To address this, I implemented my own function to change the IDs so that they are equivalent to the tip names in the data. Unfortunately, the IDs are immutable and cannot be changed.

Currently, I am attempting to create nodes myself, which I have managed to do. However, I am struggling to find a way to add those nodes to the tree. If you have a solution or potential remedy to this problem, I would love to know.

Here is my current directory where I am working: https://github.com/AyomideOjo/grasses/tree/main/Final/Enoch_O.py
My other versions are here: https://github.com/AyomideOjo/grasses/blob/main/Ayomide_O.py
and here: https://github.com/AyomideOjo/grasses/blob/main/Parham_G.py
My functional but slow cython file:https://github.com/AyomideOjo/grasses/blob/main/cython_code.pyx

Thank you for any help you may be able to offer
Ayomide Enoch Ojo

**Reply by jeromekelleher** [reply]
I see - unless you're specifically interested in using tskit I think you'd be better off using [dendropy](https://dendropy.org) to read in the newick trees.  You should be able to write a KC distance function using dendropy without too much difficulty. Tskit isn't really set-up for general phylogenetics work based on an input newick tree right now.

Using tskit would likely be faster all right though, if that's an issue.

**Reply by AyomideOjo** [reply]
Alright thank you, I gave the documentation a read its fantastic, thank you, seems like exactly what ive been looking for.
